libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1411 @@
+
+  /* Returning TRUE here, since it is not an error if the override directory
+   * does not exist */
+  return TRUE;
+}
+
+/* Returns an associative map from action-ref list to a pointer array of
+ * actions. The action-ref lists are considered to be append-only */
+GHashTable *
+eos_updater_util_flatpak_ref_actions_from_directory (const gchar   *relative_parent_path,
+                                                     GFile         *directory,
+                                                     gint           priority,
+                                                     GCancellable  *cancellable,
+                                                     GError       **error)
+{
+  g_autoptr(GHashTable) ref_actions_for_files = g_hash_table_new_full (g_str_hash,
+                                                                       g_str_equal,
+                                                                       g_free,
+                                                                       (GDestroyNotify) flatpak_remote_ref_actions_file_free);
+
+  if (!eos_updater_util_flatpak_ref_actions_append_from_directory (relative_parent_path,
+                                                                   directory,
+                                                                   ref_actions_for_files,
+                                                                   priority,
+                                                                   cancellable,
+                                                                   error))
+    return NULL;
+
+  return g_steal_pointer (&ref_actions_for_files);
+}
+
+static guint
+flatpak_ref_hash (gconstpointer data)
+{
+  return g_str_hash (flatpak_ref_get_name (FLATPAK_REF (data)));
+}
+
+static gboolean
+flatpak_ref_equal (gconstpointer a, gconstpointer b)
+{
+  return g_str_equal (flatpak_ref_get_name (FLATPAK_REF (a)),
+                      flatpak_ref_get_name (FLATPAK_REF (b)));
+}
+
+/* Squash actions on the same ref into the last action on that ref, returning
+ * a pointer array of remote ref actions, ordered by the order key in each
+ * remote ref action. */
+static GPtrArray *
+squash_ref_actions_ptr_array (GPtrArray *ref_actions)
+{
+  g_autoptr(GHashTable) hash_table = g_hash_table_new_full (flatpak_ref_hash,
+                                                            flatpak_ref_equal,
+                                                            g_object_unref,
+                                                            (GDestroyNotify) flatpak_remote_ref_action_unref);
+  GPtrArray *squashed_ref_actions = NULL;
+  gsize i;
+  gpointer key;
+  gpointer value;
+  GHashTableIter hash_iter;
+
+  for (i = 0; i < ref_actions->len; ++i)
+    {
+      FlatpakRemoteRefAction *action = g_ptr_array_index (ref_actions, i);
+      g_hash_table_replace (hash_table,
+                            g_object_ref (action->ref),
+                            flatpak_remote_ref_action_ref (action));
```

Comment:
This algorithm is going to need a bit of work for `upgrade` support, since [install X, upgrade X] squashes to [install X], and [uninstall X, upgrade X] squashes to [uninstall X].

Reply:
I'll do the work to support `upgrade` once we've worked out the rest of these review comments